feat: add abi_decode builtin (#2882)
vyperlang/vyper: docs/built-in-functions.rst

        "3233340000000000000000000000000000000000000000000000000000000000"

> .. py:function:: _abi_decode(b: Bytes, output_type: Union[typename, Tuple[typename]], unwrap_tuple: bool = True) -> Union[typename, Tuple[typename]]

Reviewer: Is this the right signature?

Reviewer 2: ```suggestion
.. py:function:: _abi_decode(b: Bytes, output_type: type_, unwrap_tuple: bool = True) -> Any
```

Reviewer: Yeah, and just be like "will return whatever decoding `b` into `output_type` gives

>     Takes a byte array as input, and returns the decoded values according to the specified output types. Used for unpacking ABIv2-encoded values.

Reviewer 2: ```suggestion
    Takes a bytestring as input, and returns the decoded values according to the specified output types. Used for unpacking ABIv2-encoded values.
```

    Once this function has seen more use we provisionally plan to put it into the ``ethereum.abi`` namespace.

    * ``b``: A byte array of a length that is between the minimum and maximum ABIv2 size bounds of the ``output type``.
    * ``output_type``: Name of the output type, or tuple of output types, to be decoded.
    * ``unwrap_tuple``: If set to True, the input is decoded as a tuple even if only one output type is specified. In other words, ``_abi_decode(b, Bytes[32])`` gets decoded as ``(Bytes[32],)``. This is the convention for ABIv2-encoded values generated by Vyper and Solidity functions. Except for very specific use cases, this should be set to True. Must be a literal.

>     Returns the decoded values.

Reviewer 2: ```suggestion
    Returns the decoded value(s), with type as specified by `output_type`.
```

    .. code-block:: python

        @external
        @view
        def foo(x: Bytes[128]) -> (uint256, Bytes[32]):
            x: uint256 = empty(uint256)
            y: Bytes[32] = empty(Bytes[32])

>             x, y =  _abi_encode(x, (uint256, Bytes[32]))

Reviewer 2: ```suggestion
            x, y =  _abi_decode(x, (uint256, Bytes[32]))
```
